SYCL/DeviceLib/imf_fp32_test.cpp

```diff
@@ -0,0 +1,1289 @@
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+
+// RUN: %clangxx -fsycl -fno-builtin -fsycl-device-lib-jit-link %s -o %t.out
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+//
+// UNSUPPORTED: cuda || hip
+
+#include <cassert>
```

Comment:
I would suggest to remove unimportant code duplicate by doing something like this:

```
#include <sycl/sycl.hpp>

template <class InputTy, class OutputTy, class FuncTy>
void test(sycl::queue &q, std::initializer_list<InputTy> Input,
          std::initializer_list<OutputTy> RefOutput, FuncTy Func,
          int Line = __builtin_LINE()) {
  auto Size = Input.size();
  assert(RefOutput.size() == Size);

  sycl::buffer<InputTy> InBuf(Size);
  {
    sycl::host_accessor InAcc(InBuf, sycl::write_only);
    int i = 0;
    for (auto x : Input)
      InAcc[i++] = x;
  }

  sycl::buffer<OutputTy> OutBuf(Size);
  q.submit([&](sycl::handler &CGH) {
     sycl::accessor InAcc(InBuf, CGH, sycl::read_only);
     sycl::accessor OutAcc(OutBuf, CGH, sycl::write_only);
     CGH.parallel_for(Size,
                      [=](sycl::id<1> Id) { OutAcc[Id] = Func(InAcc[Id]); });
   }).wait();

  sycl::host_accessor Acc(OutBuf, sycl::read_only);
  for (int i = 0; i < Size; ++i) {
    auto Expected = *(std::begin(RefOutput) + i);
    if (Expected == Acc[i])
      continue;

    std::cout << "Mismatch at line " << Line << "[" << i << "]: " << Acc[i]
              << " != " << Expected << ", input was "
              << *(std::begin(Input) + i) << std::endl;
  }
}

int foo(float x) { return x + 1; }
int foo2(float x) { return x < 5.0 ? x + 1 : x; }

// Can't pass function pointer into kernel, wrap in a lambda.
#define F(Name) [](auto x) { return (Name)(x); }

int main() {
  // Default-ctor isn't SYCL2020 compatible yet.
  sycl::queue q{sycl::default_selector{}};
  test(q, {0.0f, 1.0f}, {1, 2}, F(foo));

  {
    std::initializer_list<float> Input = {1.0f, 3.0f, 6.0f};
    test(q, Input, {2, 4, 7}, F(foo));
    test(q, Input, {2, 4, 7}, F(foo2)); // Mismatch at line 52[2]: 6 != 7, input was 6
  }
}
```

Reply:
```
// In place of
std::initializer_list<float> Input = {1.0f, 3.0f, 6.0f};

// the following
auto Input = {1.0f, 3.0f, 6.0f};

// should also work.
```

Reply:
Also, `F` can be changed into something like

```
// Maybe comma needs an extra step to properly expand
#define F(Name) [](auto x) { return (Name)(x); } , #Name
```

then the `test` could be modified to accept the function name as a parameter and report it on error as well.